Source/ui_unix/GamePad/GamePadDeviceListener.cpp

```diff
@@ -0,0 +1,230 @@
+			m_GPIEList.emplace(device.first, std::move(GamePadInput));
+		}
+nextdev:
+		continue;
+	}
+}
+
+void CGamePadDeviceListener::DisconnectInputEventCallback()
+{
+	OnInputEventCallBack = nullptr;
+	for (auto& GPI : m_GPIEList)
+	{
+		GPI.second.get()->OnInputEvent.disconnect_all_slots();
+	}
+}
+
+void CGamePadDeviceListener::UpdateDeviceList()
+{
+	std::string path = "/dev/input/";
+	for (auto & p : fs::directory_iterator(path))
+		if (p.path().filename().string().find("event") != std::string::npos)
+		{
+			AddDevice(p.path());
+		}
+}
+
+void CGamePadDeviceListener::AddDevice(const fs::path& path)
+{
+	inputdev_pair idp;
+	if(IsValidDevice(path, idp))
+	{
+		m_devicelist.insert(idp);
+		if (OnInputEventCallBack)
+		{
+			auto GamePadInput = std::make_unique<CGamePadInputEventListener>(idp.second.path, m_filter);
+			GamePadInput->OnInputEvent.connect(OnInputEventCallBack);
+			m_GPIEList.emplace(idp.first, std::move(GamePadInput));
+		}
+	}
+}
+
+void CGamePadDeviceListener::RemoveDevice(char* key)
+{
+	m_devicelist.erase(key);
+	m_GPIEList.erase(key);
+}
+
+void CGamePadDeviceListener::InputDeviceListenerThread()
+{
+	int fd = inotify_init1(IN_NONBLOCK);
+	if ( fd < 0 ) {
+		perror("inotify_init");
+	}
+
+	struct pollfd fds[2];
+	sigset_t mask;
+
+	fds[0].fd = fd;
+	fds[0].events = POLLIN;
+
+	sigemptyset(&mask);
+	sigaddset(&mask, SIGINT);
+	fds[1].fd = signalfd(-1, &mask, SFD_NONBLOCK);
+	fds[1].events = POLLIN;
+
+	sigprocmask(SIG_BLOCK, &mask, NULL);
+
+	std::string devinput_dir("/dev/input");
+	int wd = inotify_add_watch(fd, devinput_dir.c_str(), WATCH_FLAGS);
+	char buffer[EVENT_BUF_LEN];
+
+	while (m_running)
+	{
+		do
+		{
+			if (!m_running) goto cleanup;
+		} while(poll(fds, 2, 500) == 0);
+
+		int length = read(fd, buffer, EVENT_BUF_LEN);
+
+		if ( length < 0 )
+		{
+			perror("read");
+		}
+
+		for (int i=0; i < length;)
+		{
+			struct inotify_event *event = (struct inotify_event *) &buffer[i];
+			if (event->len)
+			{
+				if (std::string(event->name).find("event") != std::string::npos)
+				{
+					if (event->mask & IN_CREATE)
+					{
+						//Allow the device to be initiated
+						std::this_thread::sleep_for(std::chrono::seconds(1));
+						AddDevice(devinput_dir + "/" + event->name);
+					}
+					else if (event->mask & IN_DELETE)
+					{
+						RemoveDevice(event->name);
+					}
+				}
+			}
+			i += EVENT_SIZE + event->len;
+		}
+	}
+cleanup:
+	inotify_rm_watch(fd, wd);
+	close(fd);
+}
+
+bool CGamePadDeviceListener::IsValidDevice(const fs::path& inputdev_path, inputdev_pair& devinfo)
+{
+	if(access( inputdev_path.string().c_str(), R_OK ) == -1)
+	{
+		return false;
+	}
+	struct libevdev *dev = NULL;
+	int fd;
+	int rc = 1;
+	bool res = false;
+
+	fd = open(inputdev_path.string().c_str(), O_RDONLY);
+	if (fd < 0) {
+		libevdev_free(dev);
+		fprintf(stderr, "CGamePadDeviceListener::IsValidDevice: Error Failed to open (%s)\n", inputdev_path.string().c_str());
+		return res;
+	}
+
+	rc = libevdev_new_from_fd(fd, &dev);
```

Comment:
`rc` could be declared and initialized here since it's not used before that line.